### ISSUE TITLE
Add Homebrew Support for Instructlab CLI Setup

### ISSUE THREAD

**Issue by mohramadan911** [instructlab/instructlab#2718]
**Is your feature request related to a problem? Please describe.**

Currently, setting up the instructlab CLI requires developers to manually manage Python versions using tools like pyenv, which can be challenging, especially for those who are unfamiliar with Python environment management. This adds friction to the installation process and impacts the overall developer experience, particularly on macOS.

**Describe the solution you'd like**

I propose adding support for installing the instructlab CLI via Homebrew, a popular package manager for macOS (and Linux). By creating a Homebrew formula for the CLI, developers can easily install and manage the tool with a single command:

''' brew install instructlab '''

Homebrew will handle the version management which will simplify the setup .
Also better developer expierence as homebrew formula will create a virtual env for this setup to avoid impacting system-user machine  

**Additional context**

High level implementation steps : 

Specify the Python version as a dependency (depends_on "python@3.11").assuming 3.11 would be the stable version for most of upcoming cli releases .
Use a virtual environment (venv) to isolate the CLI and its dependencies.
Symlink the CLI binary for global usage.
Add a basic test to verify the installation.

'''brew install instructlab
instructlab --version
'''

I am happy to contribute to implementing this feature. I can:

Write the Homebrew formula and test it locally.
Provide a pull request with the formula or set up a dedicated tap repository (my-org/homebrew-instructlab).
Document the new installation process in the README file.
Let me know if this proposal aligns with the goals of the project or if there are any specific requirements to consider.

**Comment by jaideepr97** [comment]
Thanks for raising this @mohramadan911! I can see the value in this 
Also thanks for volunteering to work on it as well, please feel free to raise a PR for this if you can 🙏

**Comment by reidliu41** [comment]
I feel the same way to more focus on the `MacOS`,  I submit one [2419](https://github.com/instructlab/instructlab/pull/2419) for it with shell, but it would be very nice to setup with brew. 
And for Linux, seems shell would be better.

**Comment by ktam3** [comment]
@mohramadan911 - are you still expecting to take this on and create a PR?

**Comment by reidliu41** [comment]
Seems no response?  @ktam3 @jaideepr97 could you please assign to me? I am working on it.

**Comment by ktam3** [comment]
@reidliu41 - all set, thank you!

**Comment by reidliu41** [comment]
Hi, I did some research and tested about this, seems not a good way to way, here are my findings:
- Homebrew-core requires all the dependencies in venv that need to build it from source(unless some pkgs without source, but only accept py3-none-any)
- Based on the requirements
  - we need to handles dependencies(pkg order) by ourselves, not like, `pip` can solve it by default, we do have `180+` pkgs(and not included some compile pkgs) after installation. 
  - Download is normal(like pip do it too), but based on download, it need to compile every package, time usage of download + compile, Honestly, the installation time may longer than the simple data generate, lol, seems not a good user experience, pip would be better, faster , can cache, and already compile.
  - Compile issues are vey difficult to handle, different package different situations, not stable. I make a lot of tests, e.g. `pydantic-core` issue, `rust` compile issue, `ninja` issue, and `tqdm-multiprocess` missing `requirement-dev.txt` issue

Simple summary: `brew` is very useful and simple use for the small project or tool, since `instructLab` will become more and more stronger, and more dependencies, seems not a good way to use it.
For `MacOS` user, I think is popular and important users, so far, I would suggest using script https://github.com/instructlab/instructlab/pull/2419 (based on pip to handle), maybe a good way, and later on can easy to extend to Linux user.

May need more feedbacks, feel free to share.

@ktam3

**Comment by cdoern** [comment]
I say we either go with #2419

**Comment by mohramadan911** [comment]
I am sorry team, I was away for the last two weeks  , the PR is still on my radar .
Hi @maintainers,

I’m currently working on this issue using a forked repository under my GitHub account, as I don’t have push access to the original InstructLab repository. This setup makes it slightly more complex when testing tools like Homebrew taps or other integrations since Homebrew expects repositories to follow a strict naming convention (e.g., homebrew-<tapname>). https://docs.brew.sh/Taps#repository-naming-conventions-and-assumptions 

Specifically:

I renamed my fork to homebrew-instructlab to comply with Homebrew tap standards.
My formula (ilab.rb) resides in Formula/ilab.rb as per convention.
Testing on my fork works, but this might not directly align with your repository's structure.
Could you provide guidance or suggest how I might align this work better with the upstream repository? For example, if merging changes into a new branch here would help streamline testing or collaboration.

Thank you for your support and understanding!

**Comment by reidliu41** [comment]
I think you need need to have original repo push access. You can simply name it and test it in local.
and then your formula need to put it in `Homebrew-core` project.
Based on `Homebrew-core` requirement, you cannot simply keep align with the repository automatically, they need specific version, you need to manual maintain it in `Homebrew-core` every time if updated.
And the most challenge thing is that build from source.
Please correct me if I am misunderstanding.

**Comment by mohramadan911** [comment]
Yes @reidliu41 , can you try the below and let me know the output ? 
```sh
brew tap instructlab/ilab https://github.com/mohramadan911/homebrew-ilab
```

```sh
brew install ilab
```

you could recive linkage error between venv and brew binaries , but i tested on different users machine and it depends on the user env but in the end most of tested users were able to use the ilab with no issue so far 

this is the common error : 
```
Error: Failed changing dylib ID of /opt/homebrew/Cellar/ilab/0.22.1/instructlab/venv/lib/python3.11/site-packages/PIL/.dylibs/libpng16.16.dylib
  from /DLC/PIL/.dylibs/libpng16.16.dylib
    to /opt/homebrew/opt/ilab/instructlab/venv/lib/python3.11/site-packages/PIL/.dylibs/libpng16.16.dylib
Error: Failed to fix install linkage
The formula built, but you may encounter issues using it or linking other
formulae against it.
```

@ktam3 can you please try it as well on your machine if possible , can you please assign me back on this issue ?

**Comment by reidliu41** [comment]
@mohramadan911 
```
    # This should create virtual env
    system "python3", "-m", "venv", "--upgrade-deps", "#{install_dir}/venv"
```
Thank you so much for you hard work, actually I already tried this way, and submit a pr to homebrew-core a month ago, this way is not a allow, cannot simply use pip to install, the `url` from the `Formula`, it not just set a url in there, it need to download to local, it need to install from `source`, you need to solve all the dependencies by yourself, I would suggest reviewing the `homebrew` examples.

**Comment by mohramadan911** [comment]
@reidliu41 Hmm I can see now- very good work  , do you think does it really worth to pair on that and proceed with HomeBrew setup ? or let me know if you need any help for the shell script you are working on

**Comment by reidliu41** [comment]
@reidliu41 thanks for understanding.
1. Actually, I also shared my point a few weeks ago
https://github.com/instructlab/instructlab/issues/2718#issuecomment-2537657003
2.  The script https://github.com/instructlab/instructlab/pull/2419 also already merged
